### DISCUSSION TITLE
Multi-thread support

### DISCUSSION THREAD

**Discussion by notramo** [martenframework/marten#92] [Ideas]
Crystal supports multithreaded applications. Is it possible to enable it in Marten?

**Comment by ellmetha** [comment]
Hey! 👋 

AFAIK parallelism in Crystal is not stable and not really documented (except in a few blog posts such as [this one](https://crystal-lang.org/2019/09/06/parallelism-in-crystal/)). Though Marten is built upon the standard `HTTP::Server` abstraction, and it should (probably) be possible to enable it for a Marten project by building the server binary with `preview_mt` support.

For example:

```bash
crystal build -Dpreview_mt src/server.cr -o bin/server
CRYSTAL_WORKERS=4 bin/server
```

**Reply by notramo** [reply]
Is it possible to utilize the CPU fully, using reuseport with multiple processes, similarly to these?
https://github.com/jwoertink/lucky-cluster
https://github.com/spider-gazelle/action-controller/blob/master/clustering.md

**Reply by ellmetha** [reply]
Multi-processes is something different! Presently, there is nothing built into the framework to facilitate running multiple server processes. This may be something that will be introduced in a future release!

In the meantime, it should be noted that you can launch multiple processes of a Marten server and assign each to a different port (especially relevant if you run these server processes behind a reverse proxy such as nginx).
